From 9e536193983f023bfbb730be43c0bacebbc0cc42 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Tue, 25 Sep 2012 19:48:34 +0900
Subject: [PATCH 68/97] ext/dfbvideosink: properly read non 16-byte aligned
 buffers through MERAM

MERAM must be aligned on 16-byte boundaries. When dealing with non-16 byte
aligned buffers, start the MERAM read from the previous 16-byte boundary and
offset the VIO start address to account for the change.
---
 ext/directfb/dfbvideosink.c |   17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/ext/directfb/dfbvideosink.c b/ext/directfb/dfbvideosink.c
index a20a5b0..56f8946 100644
--- a/ext/directfb/dfbvideosink.c
+++ b/ext/directfb/dfbvideosink.c
@@ -1988,6 +1988,8 @@ gst_dfbvideosink_shvio_stretchblit (GstDfbVideoSink * dfbvideosink,
 #if defined(HAVE_SHMERAM)
   phys[SRC] = uiomux_all_virt_to_phys (src_addry);
   if (phys[SRC]) {
+    gulong py_frac = phys[SRC] & 0x0000000f;
+
     viosurface[SRC].pitch = 0;
     viosurface[SRC].bpitchy = 4096;
     viosurface[SRC].bpitchc = 4096;
@@ -1996,15 +1998,20 @@ gst_dfbvideosink_shvio_stretchblit (GstDfbVideoSink * dfbvideosink,
     meram_write_icb (dfbvideosink->meram, dfbvideosink->icby[SRC], MExxSSARA,
         phys[SRC]);
     viosurface[SRC].py =
-        (void *) meram_get_icb_address (dfbvideosink->meram,
-        dfbvideosink->icby[SRC], 0);
+        (void *) (meram_get_icb_address (dfbvideosink->meram,
+            dfbvideosink->icby[SRC], 0) + py_frac);
 
     if (dfbvideosink->icbc[SRC]) {
+      gulong physc, pc_frac;
+
+      physc = uiomux_all_virt_to_phys (src_addrc);
+      pc_frac = physc & 0x0000000f;
+
       meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[SRC],
-          MExxSSARA, uiomux_all_virt_to_phys (src_addrc));
+          MExxSSARA, physc);
       viosurface[SRC].pc = (void *)
-          meram_get_icb_address (dfbvideosink->meram, dfbvideosink->icbc[SRC],
-          0);
+          (meram_get_icb_address (dfbvideosink->meram, dfbvideosink->icbc[SRC],
+              0) + pc_frac);
     } else {
       viosurface[SRC].pc = 0;
     }
-- 
1.7.10.4

